### PR TITLE
prepare for server profiles

### PR DESCRIPTION
**Pull request by sandy081**
- make state service available in server
- Intrdouce save stragey in state service - state shall be saved immediately in server
- use UserDataProfilesService that can save profiles in server

### CODE REVIEW

File: src/vs/code/node/cliProcessMain.ts [hunk 1]
```diff
@@ -144,15 +143,14 @@ class CliMain extends Disposable {
...
-		const stateService = new StateService(environmentService, logService, fileService);
-		services.set(IStateService, stateService);
 
 		// Uri Identity
 		const uriIdentityService = new UriIdentityService(fileService);
 		services.set(IUriIdentityService, uriIdentityService);
 
 		// User Data Profiles
-		const userDataProfilesService = new UserDataProfilesService(stateService, uriIdentityService, environmentService, fileService, logService);
+		const stateService = new StateReadonlyService(SaveStrategy.DELAYED, environmentService, logService, fileService);
```
Comment (bpasero) [inline src/vs/code/node/cliProcessMain.ts:152]: I like this, I guess we were lucky before that no one did write to the state service from the CLI before when potentially an instance was already running.

However, I wonder if there is a user of state service from CLI that wants to write and is now broken, so maybe check.

File: src/vs/server/node/serverServices.ts [hunk 2]
```diff
@@ -123,14 +123,19 @@ export async function setupServerServices(connectionToken: ServerConnectionToken
...
 	// User Data Profiles
 	const userDataProfilesService = new ServerUserDataProfilesService(uriIdentityService, environmentService, fileService, logService);
 	services.set(IUserDataProfilesService, userDataProfilesService);
 
-	// Configuration
-	const configurationService = new ConfigurationService(environmentService.machineSettingsResource, fileService, new NullPolicyService(), logService);
-	services.set(IConfigurationService, configurationService);
-	await configurationService.initialize();
+	// Initialize
+	await Promise.all([
```
Comment (bpasero) [inline src/vs/server/node/serverServices.ts:135]: Should this also include the `await getMachineId()` below to run everything in parallel?

File: src/vs/platform/state/node/state.ts [hunk 3]
```diff
@@ -5,13 +5,25 @@
 
 import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
 
-export const IStateService = createDecorator<IStateService>('stateService');
-
-export interface IStateService {
+export const IStateReadonlyService = createDecorator<IStateReadonlyService>('IStateReadonlyService');
```
Comment (bpasero) [inline src/vs/platform/state/node/state.ts:8]:
```suggestion
export const IStateReadonlyService = createDecorator<IStateReadonlyService>('stateReadService');
```

File: src/vs/platform/state/node/state.ts [hunk 4]
```diff
@@ -5,13 +5,25 @@
...
+export interface IStateReadonlyService {
```
Comment (bpasero) [inline src/vs/platform/state/node/state.ts:9]: Maybe better `IStateReadService`?

File: src/vs/platform/state/node/stateService.ts [hunk 5]
```diff
@@ -10,25 +10,32 @@ import { URI } from 'vs/base/common/uri';
...
 	private closing: Promise<void> | undefined = undefined;
 
 	constructor(
 		private readonly storagePath: URI,
+		saveStrategy: SaveStrategy,
 		private readonly logService: ILogService,
-		private readonly fileService: IFileService
+		private readonly fileService: IFileService,
 	) {
+		this.flushDelayer = new ThrottledDelayer<void>(saveStrategy === SaveStrategy.IMMEDIATE ? 0 : 100 /* buffer saves over a short time */);
```
Comment (bpasero) [inline src/vs/platform/state/node/stateService.ts:38]: Should we not use any delayer in immediate mode? From the API I would have not expected a timeout here at all but sync flushing to disk.

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
Btw I only looked at the changes in state service and users, not in the larger profile related changes.

File: src/vs/code/node/cliProcessMain.ts (hunk 1 above)
Comment (sandy081) [reply src/vs/code/node/cliProcessMain.ts:152]: StateService was still readonly before except that naming was different. Readonly state service was called StateService and normal state service was called StateMainService. So we are good here

File: src/vs/server/node/serverServices.ts (hunk 2 above)
Comment (sandy081) [reply src/vs/server/node/serverServices.ts:135]: Done

File: src/vs/platform/state/node/stateService.ts (hunk 5 above)
Comment (sandy081) [reply src/vs/platform/state/node/stateService.ts:38]: Sure, done!

**Comment by sandy081** [comment]
@bpasero updated the changes and reviewing state service and its users shall be good. Thanks.

File: src/vs/code/node/cliProcessMain.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/code/node/cliProcessMain.ts:152]: Got it.

**Review by bpasero — CHANGES_REQUESTED** [review]
Two more things:
* have `state.test.ts` also test the `immediate` save mode
* should we be setting `IStateReadService` also in `main.ts` alongside `IStateService`? this would reduce the chances of breakage if code depends on the readonly service in main

**Comment by sandy081** [comment]
@bpasero - Done

**Review by bpasero — APPROVED** [review]

**Pull request merged**